File: 1210-Increasing Subsequences/1210-Increasing Subsequences.py
class Solution:
    """
    @param nums: an integer array
    @return: all the different possible increasing subsequences of the given array
    """
    def findSubsequences(self, nums):
        # Write your code here
        if len(nums) <= 1:
            return []
        path = []
        result = []
        def dfs(start):
            if len(path) >= 2:
                result.append(path[:])
            used = set()
            for i in range(start, len(nums)):
                if nums[i] not in used and (not path or nums[i] >= path[-1]):
                    path.append(nums[i])
                    used.add(nums[i])
                    dfs(i + 1)
                    path.pop()
        dfs(0)
        return result

File: 1201-1300/1210-Increasing Subsequences/1210-Increasing Subsequences.cpp
class Solution {
public:
    /**
     * @param nums: an integer array
     * @return: all the different possible increasing subsequences of the given array
     */
    vector<vector<int>> findSubsequences(vector<int> &nums) {
        // Write your code here
        vector<vector<int>> result;
        if (nums.size() == 0) {
            return result;
        }
        
        vector<int> path;
        dfs(nums, 0, path, result);
        return result;
    }
    
private:
    void dfs(vector<int>& nums, int start, vector<int>& path, vector<vector<int>>& result) {
        if (path.size() >= 2) {
            result.push_back(path);
        }
        
        unordered_set<int> table;
        for (int i = start; i < nums.size(); ++i) {
            if (table.find(nums[i]) == table.end() && (path.size() == 0 || nums[i] >= path.back())) {
                table.insert(nums[i]);
                path.push_back(nums[i]);
                dfs(nums, i + 1, path, result);
                path.pop_back();
            }
        }
    }
};
